plugins/filter_watermark/watermark.c

```diff
@@ -0,0 +1,385 @@
+    ctx->win_left_edge = 0;
+    ctx->init_flag = 0; 
+
+    /* window size */
+    str = flb_filter_get_property("window_size", f_ins);
+
+    if (str != NULL && (val = strtod(str, &endp)) > 1) {
+        ctx->window_size = val;
+    } else {
+        ctx->window_size = WINDOW_SIZE_DEFAULT_VALUE;
+    }
+    
+    /* watermark */
+    str = flb_filter_get_property("watermark", f_ins);
+
+    if (str != NULL && (val = strtod(str, &endp)) > 1) {
+        ctx->watermark = val;
+    } else {
+        ctx->watermark = WATERMARK_DEFAULT_VALUE;
+    }
+
+    /*time_field*/ 
+    str = flb_filter_get_property("time_field", f_ins);
+    if (!str) {
+        flb_plg_error(ctx->ins,  "time_field is not defined!");
+        return -1;
+    }
+    ctx->time_field = flb_sds_create(str);
+
+    return 0;
+}
+
+int deconstructor_imp(void *v) 
+{
+    struct flb_filter_watermark_record *r = v;
+    msgpack_sbuffer_free(r->sbuffer);
+    free(r);
+    return 0; 
+}
+
+int compare_imp(void *v0, void *v1) 
+{
+    struct flb_filter_watermark_record *r0 = v0;
+    struct flb_filter_watermark_record *r1 = v1;
+
+    struct tm *t0 = &(r0->time_stamp);
+    struct tm *t1 = &(r1->time_stamp);
+    time_t timer0 = mktime(t0);
+    time_t timer1 = mktime(t1);
+
+    if (timer0 < timer1)
+        return -1;
+    else if (timer0 > timer1)
+        return 1;
+    else
+        return 0;
+}
+
+static int ingest_data(struct flb_filter_watermark_ctx *context, void *ptr) 
+{
+    struct flb_filter_watermark_record *record_ptr = ptr;  
+
+    struct tm *t = &(record_ptr->time_stamp);
+    time_t tl = mktime(t);
+    const char *fmt = "%a, %d %b %Y %T %z";
+    char outstr[200] = {0};
+
+    struct c_heap_t *h = context->h;
+    time_t win_right_edge = context->win_right_edge;
+    time_t win_left_edge  = context->win_left_edge;
+    int watermark = context->watermark;
+    int window_size = context->window_size;
+
+    struct flb_filter_watermark_record *tmp_record;
+    struct tm *tmp_record_timestamp;
+  
+    int ret = 0;
+  
+    if (context->init_flag == 0) {
+        context->init_flag = 1;
+        context->win_right_edge = tl + window_size;
+        context->win_left_edge  = tl;
+        c_heap_insert(h, record_ptr);
+        return ret; 
+    }
+
+    if (tl < win_left_edge) {
+        /* Drop the data. */
+        strftime(outstr, sizeof(outstr), fmt, t);
+        flb_plg_warn(context->ins,  "Record with timestamp %s arrives too late, Drop It!", outstr);
+        ret = -1;
+        return ret;
+     } else {
+        /* Ingest the data. */
+        c_heap_insert(h, record_ptr);
+     }
+
+     /* main loop to flush out data */
+     while(tl - watermark >= win_right_edge) {
+         /*Dump all data into array to flush out*/
+         tmp_record = c_heap_read_root(h);
+         if (tmp_record == NULL) {
+             /* This place could not be reached, since at least one piece of record is in the heap*/
+             flb_plg_error(context->ins,  "Fatal Error!");
```

Comment:
Extra space?

Reply:
fixed in commit https://github.com/fluent/fluent-bit/commit/4b46f5c1048639ed37cafc345e575f6794dccba5